_rules/SC2-4-6-descriptive-labels.md

```diff
@@ -21,27 +21,38 @@ authors:
 
-- is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree), and
-- is programmatically associated with an HTML element that has one of the listed form field [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`.
+**Note**: The list of applicable [semantic roles](#semantic-roles) is derived by taking all the [ARIA 1.1](https://www.w3.org/TR/wai-aria-1.1/) roles that:
+- inherit from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and 
+- do not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.
 
-**Note**: The list of form field roles is derived by taking all the [ARIA 1.1](https://www.w3.org/TR/wai-aria-1.1/) roles that:
-- have a [semantic role](#semantic-role) that inherits from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and 
-- does not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.
-- The `option` role is not part of the list of applicable roles, because it does not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). This means [WCAG 2.1](https://www.w3.org/TR/WCAG21/) does not require it to have an accessible name.
+**Note:** The `option` role is not part of the list of applicable roles, because it has a required context role that inherits from the `select` role. Furthermore, `option` does not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). This means [WCAG 2.1](https://www.w3.org/TR/WCAG21/) does not require it to have an accessible name.
 
-**Note**: This rule is a partial check for WCAG 2.1 success criterion 2.4.6, which applies to all labels. "Label" is used in its general sense and includes text or other components with a text alternative that is presented to a user to identify a component within Web content.
+**Note**: This rule is a partial check for WCAG 2.1 success criterion 2.4.6, which applies to all labels. "Label" in WCAG is used in its general sense and includes text or other components with a text alternative that is presented to a user to identify a component within Web content.
 
-### Expectation
+### Expectation 1
 
-Each target element describes the purpose of the associated form field element.
+The `label` element, if there is any, describes the purpose of the associated form field element. The context of the element can be used to differentiate the purpose from other form fields on the same page.
+
+**Note:** In this case, context can be created by headings, fieldsets and legends, text that is near the control, etc.
```

Comment:
In the "context" definition here, I think we missed all the information giving by the Accessible Description: `title` and `aria-describedby`  (https://www.w3.org/TR/html-aam-1.0/#accessible-name-and-description-computation)

````html
<p id="shipp" aria-hidden="true">Shipping address</p>
<label>Name<input id="name" type="text" name="name" aria-describedby="shipp"/></label>
<label>Street<input id="street" type="text" name="street" aria-describedby="shipp"/></label>

<p id="bill" aria-hidden="true">Billing address</p>
<label>Name<input id="name" type="text" name="name" aria-describedby="bill"/></label>
<label>Street<input id="street" type="text" name="street" aria-describedby="bill"/></label>
````

For me, this is a passed example, but it's not if I use that rule.

Personnaly, I think the "context"/"use of the context" is a huge topic in itself and would need a seperate rule, but at least, I think a more detailled definition is needed. 

I think it is important to understand in which case the context can be used and when it can't (in particular to define, as is done on a recurrent basis in the rules, whether the context should be visible and included in the accessibility tree?) , detailled a little more what is a "text near the control" (can it be after in the source code? do it have to be visually next to the input?) ...

Reply:
@audreymaniez, I appreciate your detailed feedback :) 

Your question has got me thinking, that maybe the first expectation should be limited to cases where the `label` is visible. 

I think your point about the accessible description is very relevant, but only for Expectation 2, that is about the accessible name of the element. 

As I see it, Expectation 1 is there to handle the visible part, and here the accessible description doesn't really matter, only what can be seen.

What do you think?

Reply:
With your explanations, the difference between E1 and E2 is now more clear! For me, the only difference between Expectation 1 and 2 was the nature of the element used to give the accessible name to the form element (`label` or other methods like `aria-label` etc.).  I just now understand that by oppose `label` vs `accessible name` you certainly meant "visible label".  And I think you right, otherwise I don't really understand the necessity of two rules.

Do we scope only the `label` tag in the expectation 1? or every text near the input that could be used as a visual name ? If that is so, then I think we don't want to use the `label` in quote, because I think it is really confusing (quotes are usually used fot HTML tags or attributes).

Don't we specify it a little more? I mean, it is not obvious that by `label` we mean `visible` because a `label` tag can be not visible, as you say. 

I just suggest to replace "The `label` element" by something like "The visible label"? or the "The visible name"? 

But that was not my point, I'll try to better explain my idea :-)

First, you right, my comment concerns (now I understand well the rule) the expectation 2 only. 

Then, you write "_**Note:** In this case, context can be created by headings, fieldsets and legends, text that is near the control, etc._". 

But the context is a key element in assessing the conformity of an element, and I think it must be more detailled. The context of a form element can be its accessible description (`aria-describedby`), which is not mentionned in the "context" lists. You cited "fieldset" and "legend", but we can create form group using ARIA `role="group"` with an `aria-label` too... 

I mean, there's a variety of contexts (but also limited), and I think we need to document contexts we recognized as "usable context" as much as we can.

I was looking back to the rule for links with identical name to see if something similar on context was done but not (and I didn't notice it in that rule^^).  
Maybe my suggestion goes beyond what ACT-rules needs to specify?

I don't know if I'm more clear 😁

Reply:
For the things that were not your point, but were super relevant: 

I fixed up a bit of things both in the applicably and expectation 1 to make it more clear that it's about visible labels. 
For this rule, it's only going to consider visible labels that are actually connected to the form element (and accessible names) - at least that was the original scope of this rule. 

I will have a look at the context stuff in next round.

Reply:
The note includes "etc", because the list "headings, fieldsets and legends, text that is near the control" is not exhaustive. I would state that things like aria-describedby also add context for AT users. But we could add more things explicitly to this list. I would be fine with that.

Reply:
Now that we start digging more into the context stuff, I start to think that we will have to distinguish between what can be considered the visible context (relevant for expectation 1) and what can be considered context that included in the accessibility tree (relevant for expectation 2). 
As a sighted user, I will not have any use of descriptions created using aria-describedby, and as a screen reader user, I will not have any use of a context that has been marked up with aria-hidden="true". 

However, if we have to consider all of this, it will start to look a lot like proper definitions, which will be very hard to do, because then they have to be very strictly defined... 
I am currently a bit unsure on how to move on with this.

Reply:
If context is available in a visual way but not available for AT than that is a 1.3.1 fault and not a 2.4.6 fault and so it would not really matter I think. I think it should still pass this test then.

So the current note would be enough then, I don't think we have to distinguish between visible context and context that is included in the accessibility tree, within this rule.

Reply:
@ShadowBB, I kind of agree, but then what if there is context that is available to screen readers, but not visually? Would it still count as context?

Reply:
Good question. It stumped me for a second there. But:

If context is available for AT but not in a visual way than non-disabled users won't know what to fill into the form fields.

For example:

```
<p id="shipp" class="sr-only">Shipping address</p>
<label>Name<input id="name" type="text" name="name" aria-describedby="shipp"/></label>
<label>Street<input id="street" type="text" name="street" aria-describedby="shipp"/></label>

<p id="bill" class="sr-only">Billing address</p>
<label>Name<input id="name" type="text" name="name" aria-describedby="bill"/></label>
<label>Street<input id="street" type="text" name="street" aria-describedby="bill"/></label>
```

The above case is clearly a usability issue, not an accessibility issue. So it is not (nor should it be) a WCAG violation in my opinion.

Reply:
@ShadowBB  but back to the specific benefits of the SC 2.4.6 : 
"_Descriptive headings are especially helpful for **users who have disabilities that make reading slow and for people with limited short-term memory**. These people benefit when section titles make it possible to predict what each section contains._ "

I think the scope is broader than blind user or assistive technologies users only, and hidden text give an advantage to blind user, and disadvantage disabled users that do not rely on an AT.

Reply:
> 
> 
> @ShadowBB but back to the specific benefits of the SC 2.4.6 :
> "_Descriptive headings are especially helpful for **users who have disabilities that make reading slow and for people with limited short-term memory**. These people benefit when section titles make it possible to predict what each section contains._ "
> 
> I think the scope is broader than blind user or assistive technologies users only, and hidden text give an advantage to blind user, and disadvantage disabled users that do not rely on an AT.

I agree with the fact that the scope is broader than blind users or assistive technology users. However hidden text doesn't give a disadvantage to disabled users that do not rely on an AT _when compared to non-disabled users_.

My example harms all users (apart from those that use screen readers I guess), not just users with disabilities.

Reply:
### Expectation 1: 
...

**Note:** In this case, context can be created by headings, fieldsets and legends, text that is near the control, etc. Only [visible](#visible) elements should be considered as context for this expectation. 

### Expectation 2
...

**Note:** In this case, context can be created by headings, fieldsets and legends, text that is near the control, etc. For this expectation elements that are [included in the accessibility tree](#included-in-the-accessibility-tree) should be considered for context as well, e.g. elements connected through `aria-describedby`, or `role="group"` with an [accessible name](#accessible-name).

**Note:** If a context that is necessary for differentiating form fields from each other is [visible](#visible), but not [included in the accessibility tree](#included-in-the-accessibility-tree), this might be a violation under other [WCAG](https://www.w3.org/TR/WCAG21/) success criteria (e.g. 1.3.1 Info and Relationships), but this is not considered a failing condition for this rule. For this rule, it is enough that the context is there in some form.

Reply:
I think this captures it correctly! Nicely done!

I see no flaw in the wording or grammar.

Reply:
Great work @annethyme :-) I think the two expectations are now more understandable :+1:

Reply:
@audreymaniez, so glad to hear that! 
If you have time to review again, it would be much appreciated.